### PR TITLE
[AD-330] Implement wallet restore dialog

### PR DESCRIPTION
**Pull request by maksbotan**
**YT issue:** https://issues.serokell.io/issue/AD-330

**Checklist:**

- [x] Updated docs if necessary
  - [x] [README](README.md)
  - [x] [TUI usage guide](docs/usage-tui.md)
- [x] Adressed HLint warnings and hints
- [x] Rebased branch on current master and squashed all "Address PR comment" commits
- [x] Tested my changes if they modify code

**Description:**
Added QComboBox with two options: create new wallet and import.

### CODE REVIEW

File: ui/qt-lib/src/Ariadne/UI/Qt/Widgets/Dialogs/NewWallet.hs [hunk 1]
```diff
@@ -212,21 +267,37 @@ hasPasswordToggled nw@NewWallet{..} checked = do
...
+    when nwHasPassword $
+      guard $ not (null nwPasswordCandidate) && nwPasswordCandidate == nwRepeatPassword
+    let nwPassword = if nwHasPassword then Just nwPasswordCandidate else Nothing
+
+    nwSpecifier <- case mode of
+      CreateWallet -> do
+        return $ NewWalletName
+      ImportWallet -> do
+        guard $ not $ null nwMnemonic
+        guard $ length (words nwMnemonic) >= 12
```
Comment (maksbotan) [inline ui/qt-lib/src/Ariadne/UI/Qt/Widgets/Dialogs/NewWallet.hs]: Not sure if this check is needed here. What do you think?

Comment (gromakovsky) [reply ui/qt-lib/src/Ariadne/UI/Qt/Widgets/Dialogs/NewWallet.hs]: I think it won't harm.

File: ui/qt-lib/src/Ariadne/UI/Qt/Widgets/Dialogs/NewWallet.hs [hunk 2]
```diff
@@ -128,13 +166,35 @@ initNewWallet = do
...
   addRow layout walletNameLabel walletName
+
+  walletMnemonicLabel <- QLabel.newWithText ("<b>MNEMONIC PHRASE</b>" :: String)
+  walletMnemonic <- QLineEdit.new
+
+  (walletMnemonicWidget, walletMnemonicLayout) <- createSubWidget
+  addRow walletMnemonicLayout walletMnemonicLabel walletMnemonic
+
+  fullRestoreLabel <- QLabel.newWithText ("Perform full restore" :: String)
+  fullRestore <- QCheckBox.new
```
Comment (gromakovsky) [inline ui/qt-lib/src/Ariadne/UI/Qt/Widgets/Dialogs/NewWallet.hs]: I think it should be enabled by default (like in TUI).

File: ui/qt-app/Glue.hs [hunk 3]
```diff
@@ -116,6 +116,15 @@ knitFaceToUI UiFace{..} KnitFace{..} =
...
             ++ maybe [] ((:[]) . Knit.ArgKw "pass" . Knit.ExprLit . Knit.toLit . Knit.LitString) password
           )
+      UiRestoreWallet name password mnemonic full -> do
+        Right $ Knit.ExprProcCall
+          (Knit.ProcCall Knit.restoreCommandName $
+            [ Knit.ArgKw "name" . Knit.ExprLit . Knit.toLit . Knit.LitString $ name
+            , Knit.ArgKw "mnemonic" . Knit.ExprLit . Knit.toLit . Knit.LitString $ mnemonic
+            , Knit.ArgKw "full" . Knit.componentInflate . Knit.ValueBool $ full
+            ]
+            ++ maybe [] ((:[]) . Knit.ArgKw "pass" . Knit.ExprLit . Knit.toLit . Knit.LitString) password
```
[the remark this answers is not in the document]
Comment (maksbotan) [reply ui/qt-app/Glue.hs]: This is copy-pasted from `UiNewWallet` part :unamused:

File: ui/qt-app/Glue.hs [hunk 4]
```diff
@@ -136,6 +145,9 @@ knitFaceToUI UiFace{..} KnitFace{..} =
           fromResult result >>= fromValue >>= \case
             Knit.ValueList l -> Right [s | Just (Knit.ValueString s) <- Knit.fromValue <$> l]
             _ -> Left "Unrecognized return value"
+      UiRestoreWallet {} ->
+        Just . UiRestoreWalletCommandResult . either UiRestoreWalletCommandFailure (const UiRestoreWalletCommandSuccess) $
```
[the remark this answers is not in the document]
Comment (maksbotan) [reply ui/qt-app/Glue.hs]: This is copy-pasted as well. I trimmed all such lines to 100 symbols max.

File: stack.yaml [hunk 5]
```diff
@@ -147,7 +147,7 @@ packages:
 ## We are using our fork because we added some stuff there
 - location:
     git: https://github.com/serokell/qtah.git
-    commit: da34829dfa1906f26e5bba7de0fde6ba577951cb
+    commit: 823e9d884996b4873a3c841d8662fc66d2c95168
```
[the remark this answers is not in the document]
Comment (maksbotan) [reply stack.yaml]: I edited first commit's message and force-pushed.

File: ui/qt-lib/src/Ariadne/UI/Qt/Widgets/Dialogs/NewWallet.hs (hunk 2 above)
Comment (maksbotan) [reply ui/qt-lib/src/Ariadne/UI/Qt/Widgets/Dialogs/NewWallet.hs]: Done.

**Pull request merged**